docs/architecture/adr-020-deterministic-square-construction.md

```diff
@@ -0,0 +1,82 @@
+```
+
+If there is sufficient remaining capacity the `CompactShareCounter` will record the share and return true. The `remainder` field is used to record the remaining bytes in the ultimate share.
+This emulates exactly how the encoding is done resulting in no error to the actual shares used.
+
+For sparse shares i.e. the blobs themselves, the `SparseShareCounter` estimates the worst case intra-blob padding. This is according to the rules specified in [ADR013](./adr-013-non-interactive-default-rules-for-zero-padding.md). To paraphrase, the padding between blobs is dictated based on what would be the first subtree root in the merkle mountain range over the set of shares that correspond to that blob. The higher the first subtree root, the greater the worst case padding could possibly be. As an example, say a blob consisted of 11 shares, and that according to the specification in the construction of a `ShareCommitment`, the subtree roots (i.e. the merkle mountain range was): 4, 4, 2, 1. The worst case padding is 4 - 1 = 3 shares. Accordingly, as any blob with shares less than or equal to `SubtreeRootThreshold` has a subtree root height of 1, these blobs will all have zero padding. The `SpareShareCounter` is thus not optimal in it's estimation of the shares required as padding might be less than what was estimated, however further investigation is left for later discussion. Note that as we are taking the worst case padding, we don't need to worry about the order of the blobs. They will be ordered by namespace only at the end.
+
+With these two structs, we can safely guarantee that all staged transactions can be included in the square and from the estimation can formulate the minimum square size that will be used to calculate the share index for all the blobs. The construction of the square itself remains the same. Additionally, as the share index and square size can be deterministically calculated it no longer needs to be gossiped to all nodes in the consensus network. However, for verifiability purposes the square size should still remain in the `Data` struct.
+
+Both `PrepareProposal` and `ProcessProposal` will as a result, call much the same methods. Verification is thus reduced to: did I create the same square as you, rather than is your version of the square valid. The main difference is that `PrepareProposal` will handle overflow of transactions by discarding them, while `ProcessProposal` will handle overflow by rejecting the block.
+
+## Consequences
+
+### Positive
+
+- By staging transactions in order of priority, we prevent the removal of a higher priority transaction that had a higher namespace than a transaction with a lower namespace and lower priority. [#1519](https://github.com/celestiaorg/celestia-app/issues/1519)
+- Block proposers aren't able to spam the network as they no longer have control over the size of the square insofar as the amount of transactions included.
```

Comment:
[not blocking]
is this referring to proposers building large padded empty squares?

> aren't able to spam the network

not blocking, but we might be able to rephrase this to be a tad clearer. They still could spam the network with non-sense 0 fee txs, correct?